src/charm.py

```diff
@@ -605,38 +666,47 @@ def _on_nrpe_targets_changed(self, event: Optional[NrpeTargetsChangedEvent]):
                 label_rules=False,
             )
 
-    def _set_status(self):
-        messages = []
-        if (
-            (self._stored.have_grafana or self._stored.have_gagent)
-            and not self._stored.have_dashboards
-        ) or (  # pyright: ignore
-            self._stored.have_dashboards
-            and not (self._stored.have_grafana or self._stored.have_gagent)  # pyright: ignore
-        ):
-            messages.append("one of (Grafana|dashboard|grafana-agent)")
+    def _set_status(self, _event):
+        # Put charm in blocked status if all incoming relations are missing
+        # We could obtain the set of active relations with:
+        # {k for k, v in self.model.relations.items() if v}
+        # but that would incur a relation-list and multiple relation-get calls, so building up from
+        # stored state instead.
+        active_relations = {
+            rel
+            for (rel, indicator) in [
+                ("downstream-grafana-dashboard", self._stored.have_grafana),
+                ("cos-agent", self._stored.have_gagent),
+                ("dashboards", self._stored.have_dashboards),
+                ("downstream-logging", self._stored.have_loki),
+                ("filebeat", self._stored.have_filebeat),
+                ("monitors", self._stored.have_nrpe),
+                ("general-info", self._stored.have_general_info_nrpe),
+                ("downstream-prometheus-scrape", self._stored.have_prometheus),
+                ("prometheus-target", self._stored.have_targets),
+                ("prometheus-rules", self._stored.have_prometheus_rules),
+                ("prometheus", self._stored.have_prometheus_manual),
+            ]
+            if indicator
+        }
+
+        # Set blocked if _all_ incoming relations are missing. This helps notice under-configured
+        # or redundant cos-proxy instances.
+        if not set(self.relation_pairs.keys()).intersection(active_relations):
+            self.unit.status = BlockedStatus("Missing incoming relation(s)")
```

Comment:
I found this a little clearer before being updated.  When debugging [this conversation](https://github.com/canonical/cos-proxy-operator/pull/145/files/339cbfbe315a94009b7ad585acb7bfc0279abdf5#r1649170085), I misunderstood this error as the charm requring a specific incoming relation rather than saying I needed one *of any* supported incoming relations. Maybe reword to
```suggestion
            self.unit.status = BlockedStatus("Charm requires at least one incoming relation")
```

or similar?  we could list the relations that satisfy this if we want, or just say it is missing.

Reply:
Done in next commit.